core/communications/httpCommunication.go

```diff
@@ -1054,145 +1417,349 @@ func (communication *HTTP) handlePutData(orgID string, objectType string, object
 			if _, err := io.Copy(writer, dataReader); err != nil {
+				hasError = true
 				SendErrorResponse(writer, err, "", 0)
 			}
 			if err := Store.CloseDataReader(dataReader); err != nil {
+				hasError = true
 				SendErrorResponse(writer, err, "", 0)
 			}
 			if trace.IsLogging(logger.DEBUG) {
 				trace.Debug("Handling object get data, update notification for %s %s %s %s, status: %s\n", objectType, objectID, destType, destID, common.Data)
 			}
-			// update notification only if current notification.InstanceID == metadata.InstanceID && current notification.status == "updated"
-			if updateNotificationRecord {
+		}
+	} else {
+		// dataLength is partial && no range header
+		if objectData, eof, length, err := Store.ReadObjectData(orgID, objectType, objectID, dataLength, startOffset); err != nil {
+			SendErrorResponse(writer, err, "", 0)
+		} else {
+			if len(objectData) == 0 {
 				if trace.IsLogging(logger.DEBUG) {
-					trace.Debug("Handling object get data, update notification status to data for %s %s %s %s %s\n", orgID, objectType, objectID, destType, destID)
-				}
-				notification := common.Notification{ObjectID: objectID, ObjectType: objectType,
-					DestOrgID: orgID, DestID: destID, DestType: destType, Status: common.Data, InstanceID: instanceID, DataID: dataID}
-				if err = Store.UpdateNotificationRecord(notification); err != nil {
-					if log.IsLogging(logger.ERROR) {
-						log.Error("Handling object get data, failed to update notification for %s %s %s %s with status: %s\n", objectType, objectID, destType, destID, common.Data)
-					}
-				} else {
-					if trace.IsLogging(logger.DEBUG) {
-						log.Debug("Handling object get data, update notification for %s %s %s %s with status %s is done\n", objectType, objectID, destType, destID, common.Data)
-					}
+					trace.Debug("Object data length is 0 for %s %s, return 404", objectType, objectID)
 				}
+				writer.WriteHeader(http.StatusNotFound)
 			} else {
-				if trace.IsLogging(logger.DEBUG) {
-					trace.Debug("Handling object get data, return without update notification status to data for %s %s %s %s %s, set updateNotificationRecord to %t \n", orgID, objectType, objectID, destType, destID, updateNotificationRecord)
+				dataReader := bytes.NewReader(objectData)
+				writer.Header().Add("Content-Type", "application/octet-stream")
+				writer.Header().Add("Content-Length", strconv.Itoa(length))
+				if eof {
+					endOffset = objectMeta.ObjectSize - 1
+				}
+				writer.Header().Add("Content-Range", fmt.Sprintf("bytes %d-%d/%d", startOffset, endOffset, objectMeta.ObjectSize))
+				writer.WriteHeader(http.StatusPartialContent)
+
+				if _, err := io.Copy(writer, dataReader); err != nil {
+					hasError = true
+					SendErrorResponse(writer, err, "", 0)
+				}
+				if err := Store.CloseDataReader(dataReader); err != nil {
+					hasError = true
+					SendErrorResponse(writer, err, "", 0)
 				}
 			}
 		}
 	}
+
+	if hasError == false && trace.IsLogging(logger.DEBUG) {
```

Comment:
I think you can delete this section for the log message. Since we will always be using chunks, the download time isn't as important as it would have been if the whole object  was streamed